### PR TITLE
Ron azar Resources and Services

### PR DESCRIPTION
**Pull request by RonAzar**

### CODE REVIEW

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    @GET
+    @Path("getAllOwners/{companyId}")
+    fun getAllOwners(@PathParam("companyId") companyId: Long): Response {
+        if (companyId <= 0) {
+            return Response.status(Response.Status.BAD_REQUEST).entity("Invalid companyId provided").build()
+        }
+        val owners = ownerService.getAllOwners(companyId)
+        if (owners.isEmpty()) {
+            return Response.status(Response.Status.NOT_FOUND).entity("No owners found for the given company").build()
+        }
```
Comment (RonAzar) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt:25]: Move to service

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    @POST
+    @Path("insertNewOwner")
+    fun insertNewOwner(newOwner: OwnerRecord): Response {
+        ownerService.insertOwner(newOwner)
+        return Response.status(Response.Status.CREATED).entity("Owner successfully inserted").build()
+    }
+
+    @GET
+    @Path("getOwnerByPetId/{petId}/companyId/{companyId}")
+    @Produces(MediaType.APPLICATION_JSON)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt:31]: You already have `@Produces` in the begging of the class, this is redundant.

Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt:31]: Why 2 twice companyId in path?

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    fun getOwnerByPetId(
+        @PathParam("petId") petId: Long,
+        @PathParam("companyId") companyId: Long
+    ): Response {
+        val owner = ownerService.getOwnerByPetId(petId, companyId)
+        return Response.ok(owner).build()
+    }
+
+    @GET
+    @Path("getOwnerByEmployeeId/{employeeId}/CompanyId/{companyId}")
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt:41]: Why 2 twice companyId in path?

File: src/main/kotlin/com/hibob/academy/resource/PetsResource.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import jakarta.ws.rs.core.Response
+import org.springframework.stereotype.Controller
+
+@Controller  // Marks the class as a controller, meaning it's responsible for handling incoming HTTP requests
+@Path("/api/ron/azar/pets/")  // Defines the base URL path that this controller will handle
+@Produces(MediaType.APPLICATION_JSON)  // Specifies that the responses produced by this controller will be in JSON format
+@Consumes(MediaType.APPLICATION_JSON)  // Specifies that this endpoint accepts JSON input
+class PetsResource (private val petService: PetService){
+    @GET
+    @Path("getAllPetsByType/{petType}/companyId/{companyId}")
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetsResource.kt:23]: Why 2 twice companyId in path?

File: src/main/kotlin/com/hibob/academy/resource/PetsResource.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    @GET
+    @Path("getAllPets/{companyId}")
+    fun getAllPets(@PathParam("companyId") companyId: Long): Response {
+        val petsList = petService.getAllPets(companyId)
+        return Response.ok(petsList).build()
+    }
+
+    @GET
+    @Path("getPetById/{petId}/companyId/{companyId}")
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetsResource.kt:37]: Why 2 twice companyId in path?

File: src/main/kotlin/com/hibob/academy/resource/PetsResource.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    fun getPetById(@PathParam("petId") petId: Long, @PathParam("companyId") companyId: Long): Response {
+        val pet = petService.getPetById(petId, companyId)
+        return Response.ok(pet).build()
+    }
+
+    @POST
+    @Path("insertNewPet")
+    fun insertNewPet(newPet: PetRecord): Response {
+        petService.insertNewPet(newPet)
+        return Response.status(Response.Status.CREATED).entity("Pet successfully inserted").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetsResource.kt:47]: Response.ok is enough

File: src/main/kotlin/com/hibob/academy/resource/PetsResource.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @PUT
+    @Path("updatePetOwnerId/{petId}/newOwnerId/{newOwnerId}/companyId/{companyId}")
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetsResource.kt:51]: No need to specify the next param

File: src/main/kotlin/com/hibob/academy/service/OwnerService.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        if (newOwnerId < 0L)
+        {
+            throw IllegalArgumentException("Owner insertion failed...")
+        }
+        return newOwnerId
+    }
+    fun getOwnerByEmployeeIdAndCompanyId(employeeId: String, companyId: Long): OwnerData{
+        val owner = ownerDao.getOwnerByEmployeeIdAndCompanyId(employeeId, companyId) ?: throw NoSuchElementException("No owner with id $employeeId found")
+        return owner
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/service/OwnerService.kt:22]: enter between functions

File: src/main/kotlin/com/hibob/academy/service/PetService.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        if (pet.ownerId != null) {
+            throw IllegalArgumentException("Pet already has an owner")
+        }
+
+        // Update the ownerId if it is currently null
+        val rowsAffected = petDao.updatePetOwnerId(petId, petOwnerId, companyId)
+        if (rowsAffected == 0) {
+            throw IllegalArgumentException("Failed to update pet's owner. The pet may already have an owner.")
+        }
+        return "Pet owner ID updated successfully"
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/service/PetService.kt:46]: What we will do with this String?
200 response is enough

**Pull request merged**